### PR TITLE
[Fix] Overlapping stored flows 

### LINKED ISSUE #23: Duplicated flows are created when using overlapping requests
Hi,

If we send a request to flow_manager to create overlapping flows, flow_manager will save both flows on storehouse and the consistency check routine will keep inserting and removing flows on each run.

Steps to reproduce:
1. Run kytos with flow_manager napp from master branch, and using -E
2. Request the initial flows:
```
curl -d '{"flows": [{"priority": 10, "match": {"in_port": 1, "dl_vlan": 100}, "actions": [{"action_type": "output", "port": 2}]}, {"priority": 10, "match": {"in_port": 2, "dl_vlan": 200}, "actions": [{"action_type": "output", "port": 3}]}, {"priority": 10, "match": {"in_port": 3, "dl_vlan": 300}, "actions": [{"action_type": "output", "port": 4}]}]}' -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01
```
3. show the flows installed in the switches flow_table and saved in the storehouse
```
# ovs-ofctl dump-flows s1
 cookie=0x0, duration=171.878s, table=0, n_packets=116, n_bytes=4872, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
 cookie=0x0, duration=10.031s, table=0, n_packets=0, n_bytes=0, priority=10,in_port="s1-eth1",dl_vlan=100 actions=output:"s1-eth2"
 cookie=0x0, duration=10.019s, table=0, n_packets=0, n_bytes=0, priority=10,in_port="s1-eth2",dl_vlan=200 actions=output:"s1-eth3"
 cookie=0x0, duration=10.009s, table=0, n_packets=0, n_bytes=0, priority=10,in_port="s1-eth3",dl_vlan=300 actions=output:"s1-eth4"
# curl http://127.0.0.1:8181/api/kytos/storehouse/v1/kytos.flow.persistence/af2fc8cae1aa4860a20eff290a5c9c8a
{"flow_persistence":{
"00:00:00:00:00:00:00:01":{
    "flow_list":[
        {"command":"add","flow":{"actions":[{"action_type":"output","port":4294967293}],"match":{"dl_type":35020,"dl_vlan":3799},"priority":1000,"table_id":0}},
        {"command":"add","flow":{"actions":[{"action_type":"output","port":2}],"match":{"dl_vlan":100,"in_port":1},"priority":10}},
        {"command":"add","flow":{"actions":[{"action_type":"output","port":3}],"match":{"dl_vlan":200,"in_port":2},"priority":10}},
        {"command":"add","flow":{"actions":[{"action_type":"output","port":4}],"match":{"dl_vlan":300,"in_port":3},"priority":10}}
   ]},
"00:00:00:00:00:00:00:02":{
    "flow_list":[
        {"command":"add","flow":{"actions":[{"action_type":"output","port":4294967293}],"match":{"dl_type":35020,"dl_vlan":3799},"priority":1000,"table_id":0}}
   ]},
"00:00:00:00:00:00:00:03":{
   "flow_list":[
        {"command":"add","flow":{"actions":[{"action_type":"output","port":4294967293}],"match":{"dl_type":35020,"dl_vlan":3799},"priority":1000,"table_id":0}}
   ]}
}}
```
4. Request an overlapping flow:
```
curl -d '{"flows": [{"priority": 10, "match": {"in_port": 1, "dl_vlan": 100} } ]}' -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01
```
5. check the saved storehouse
```
{"flow_persistence":{
"00:00:00:00:00:00:00:01":{
   "flow_list":[
       {"command":"add","flow":{"actions":[{"action_type":"output","port":4294967293}],"match":{"dl_type":35020,"dl_vlan":3799},"priority":1000,"table_id":0}},
       {"command":"add","flow":{"actions":[{"action_type":"output","port":2}],"match":{"dl_vlan":100,"in_port":1},"priority":10}},
       {"command":"add","flow":{"actions":[{"action_type":"output","port":3}],"match":{"dl_vlan":200,"in_port":2},"priority":10}},
       {"command":"add","flow":{"actions":[{"action_type":"output","port":4}],"match":{"dl_vlan":300,"in_port":3},"priority":10}},
       {"command":"add","flow":{"match":{"dl_vlan":100,"in_port":1},"priority":10}}
   ]},
"00:00:00:00:00:00:00:02":{
   "flow_list":[
       {"command":"add","flow":{"actions":[{"action_type":"output","port":4294967293}],"match":{"dl_type":35020,"dl_vlan":3799},"priority":1000,"table_id":0}}
    ]},
"00:00:00:00:00:00:00:03":{
   "flow_list":[
       {"command":"add","flow":{"actions":[{"action_type":"output","port":4294967293}],"match":{"dl_type":35020,"dl_vlan":3799},"priority":1000,"table_id":0}}
    ]}
}}
```
6. check the logs
```
Aug 27 15:56:45 39e27053c9b7 werkzeug:INFO _internal:113:  127.0.0.1 - - [27/Aug/2021 15:56:45] "#033[37mPOST /api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01 HTTP/1.1#033[0m" 202 -
Aug 27 15:56:45 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO storehouse:114:  Flow saved in kytos.flow.persistence.2cd3af668f974e7cbd6ab1d145a91dfd
Aug 27 15:56:45 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO storehouse:114:  Flow saved in kytos.flow.persistence.2cd3af668f974e7cbd6ab1d145a91dfd
Aug 27 15:56:45 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO storehouse:114:  Flow saved in kytos.flow.persistence.2cd3af668f974e7cbd6ab1d145a91dfd
Aug 27 15:56:55 39e27053c9b7 werkzeug:INFO _internal:113:  127.0.0.1 - - [27/Aug/2021 15:56:55] "#033[37mPOST /api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01 HTTP/1.1#033[0m" 202 -
Aug 27 15:56:55 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO storehouse:114:  Flow saved in kytos.flow.persistence.2cd3af668f974e7cbd6ab1d145a91dfd
Aug 27 15:56:56 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO main:197:  A consistency problem was detected in switch 00:00:00:00:00:00:00:01.
Aug 27 15:56:56 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO main:201:  Flow forwarded to switch 00:00:00:00:00:00:00:01 to be installed. Flow: {'flows': [{'priority': 10, 'actions': [{'port': 2, 'action_type': 'output'}], 'match': {'dl_vlan': 100, 'in_port': 1}}]}
Aug 27 15:56:59 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO main:197:  A consistency problem was detected in switch 00:00:00:00:00:00:00:01.
Aug 27 15:56:59 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO main:201:  Flow forwarded to switch 00:00:00:00:00:00:00:01 to be installed. Flow: {'flows': [{'priority': 10, 'match': {'dl_vlan': 100, 'in_port': 1}}]}
Aug 27 15:57:02 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO main:197:  A consistency problem was detected in switch 00:00:00:00:00:00:00:01.
Aug 27 15:57:02 39e27053c9b7 kytos.napps.kytos/flow_manager:INFO main:201:  Flow forwarded to switch 00:00:00:00:00:00:00:01 to be installed. Flow: {'flows': [{'priority': 10, 'actions': [{'port': 2, 'action_type': 'output'}], 'match': {'dl_vlan': 100, 'in_port': 1}}]}
```

### PR DESCRIPTION
**Pull request by viniarck**
Fixes #23 that @italovalcy has reported 

- This PR is same as PR #42 (but I'm using the `fix/overlapping_flows` branch directly on the upstream repo to facilitate visualizing diffs since there are upcoming PRs that will depend on this branch here, let's try this out to see if it facilitates for us)

### Description of the change

- Augmented ``_add_flow_store`` to overwrite overlapping flows
- Added unit test to cover it

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -326,11 +325,23 @@ def _add_flow_store(self, flow_dict, switch):
...
+                    (
+                        stored_flows[i]["flow"].get("priority", 0)
+                        == flow_dict.get("priority", 0),
+                        match_flow(flow_dict, version, stored_flows[i]["flow"]),
+                    )
+                ):
+                    stored_flows_box[switch.id][cookie][i] = installed_flow
+                    break
+            else:
+                stored_flows_box[switch.id][cookie].append(installed_flow)
```
Comment (ajoaoff) [inline main.py:344]: Why `enumerate` and then using just the index?

Comment (viniarck) [reply main.py]: Good catch. I meant to also use the `v` value (`stored_flow`) this way here https://github.com/kytos-ng/flow_manager/pull/44/commits/5b667e9290a10ed4c95f2e991570a84f99defb12, I've just updated this.

File: tests/unit/test_main.py [hunk 2]
```diff
@@ -405,6 +408,89 @@ def test_check_switch_consistency_add(self, *args):
         self.napp.check_switch_consistency(switch)
         mock_install_flows.assert_called()
 
+    @patch("napps.kytos.flow_manager.storehouse.StoreHouse.save_flow")
+    def test_add_overlapping_flow(self, *args):
```
Comment (italovalcy) [inline tests/unit/test_main.py:412]: @viniarck if you think it is useful, I would suggest adding another unit test for the example I've provided in the other comment, so we can validate the strict/non-strict matching.

**Review by italovalcy — CHANGES_REQUESTED** [review]
This PR is very interesting and the solution is pretty robust and straightforward. I just made two comments concerning the strict/non-strict match trying to collaborate with the proposal.

File: main.py [hunk 3]
```diff
@@ -326,11 +325,23 @@ def _add_flow_store(self, flow_dict, switch):
...
         else:
-            stored_flows_box[switch.id][cookie].append(installed_flow)
+            version = switch.connection.protocol.version
+            stored_flows = stored_flows_box[switch.id].get(cookie, [])
+            for i, stored_flow in enumerate(stored_flows):
+                if all(
+                    (
+                        stored_flow["flow"].get("priority", 0)
+                        == flow_dict.get("priority", 0),
+                        match_flow(flow_dict, version, stored_flow["flow"]),
```
[the remark this answers is not in the document]
Comment (viniarck) [reply main.py]: Good catch @italovalcy, really appreciated your help and review, I'm glad this was caught before getting merged. I've just pushed a few commits, I took the chance to also extract some functions, see these commits  https://github.com/kytos-ng/flow_manager/pull/44/commits/b8fc4e55409292acdbd6342b0656beeb3c289dfe and https://github.com/kytos-ng/flow_manager/pull/44/commits/2fe09ea1367576346447037d7a04317ac11be760.

File: tests/unit/test_main.py (hunk 2 above)
Comment (viniarck) [reply tests/unit/test_main.py:412]: for sure, @italovalcy, I've added this test case on this commit https://github.com/kytos-ng/flow_manager/pull/44/commits/59b155f7d9c87e3028b777aa38dcc6d28164cc42 based on the example provided, I slightly changed it though to start with 4 flows installed, but same semantics, check it out, and let me know what you think and if you could do a second pass review, thanks. 

I also took the chance to run it on mininet the same example you gave to double check, it worked: 

```
{
    "flow_persistence": {
        "00:00:00:00:00:00:00:01": {
            "0": [
                {
                    "created_at": "2021-12-02T15:04:17",
                    "flow": {
                        "actions": [
                            {
                                "action_type": "output",
                                "port": 4294967293
                            }
                        ],
                        "match": {
                            "dl_type": 35020,
                            "dl_vlan": 3799
                        },
                        "priority": 1000,
                        "table_id": 0
                    }
                },
                {
                    "created_at": "2021-12-02T15:14:16",
                    "flow": {
                        "actions": [
                            {
                                "action_type": "output",
                                "port": 2
                            }
                        ],
                        "match": {
                            "dl_vlan": 100,
                            "in_port": 1
                        },
                        "priority": 10
                    }
                },
                {
                    "created_at": "2021-12-02T15:14:16",
                    "flow": {
                        "actions": [
                            {
                                "action_type": "output",
                                "port": 3
                            }
                        ],
                        "match": {
                            "dl_vlan": 200,
                            "in_port": 1
                        },
                        "priority": 10
                    }
                },
                {
                    "created_at": "2021-12-02T15:14:16",
                    "flow": {
                        "actions": [
                            {
                                "action_type": "output",
                                "port": 4
                            }
                        ],
                        "match": {
                            "dl_vlan": 300,
                            "in_port": 1
                        },
                        "priority": 10
                    }
                },
                {
                    "created_at": "2021-12-02T15:14:41",
                    "flow": {
                        "match": {
                            "in_port": 1
                        },
                        "priority": 10
                    }
                }
            ]
        },
        "00:00:00:00:00:00:00:02": {
            "0": [
                {
                    "created_at": "2021-12-02T15:04:17",
                    "flow": {
                        "actions": [
                            {
                                "action_type": "output",
                                "port": 4294967293
                            }
                        ],
                        "match": {
                            "dl_type": 35020,
                            "dl_vlan": 3799
                        },
                        "priority": 1000,
                        "table_id": 0
                    }
                }
            ]
        },
        "00:00:00:00:00:00:00:03": {
            "0": [
                {
                    "created_at": "2021-12-02T15:04:17",
                    "flow": {
                        "actions": [
                            {
                                "action_type": "output",
                                "port": 4294967293
                            }
                        ],
                        "match": {
                            "dl_type": 35020,
                            "dl_vlan": 3799
                        },
                        "priority": 1000,
                        "table_id": 0
                    }
                }
            ]
        }
    }
}```

File: match.py [hunk 4]
```diff
@@ -105,17 +116,34 @@ def match10_no_strict(flow_dict, args):
...
     cookie = flow_to_install.get("cookie", 0) & flow_to_install.get("cookie_mask", 0)
     cookie_stored = stored_flow_dict.get("cookie", 0) & flow_to_install.get(
         "cookie_mask", 0
     )
     if cookie and cookie != cookie_stored:
         return False
+    return True
+
+
+def _match_keys(flow_to_install, stored_flow_dict, flow_to_install_keys):
```
Comment (viniarck) [inline match.py:130]: This was extracted to be re-used, but to add support for masked values we'll cover on this issue #24 in the future.

**Review by italovalcy — APPROVED** [review]

**Comment by viniarck** [comment]
thanks for reviewing, and the second pass Italo and Antonio.

**Pull request merged**